src/cascadia/TerminalControl/TermControl.cpp

```diff
@@ -227,34 +253,174 @@ namespace winrt::Microsoft::Terminal::TerminalControl::implementation
+        auto weakThis{ this->get_weak() };
 
-        const Search::Sensitivity sensitivity = caseSensitive ?
-                                                    Search::Sensitivity::CaseSensitive :
-                                                    Search::Sensitivity::CaseInsensitive;
+        // If no matches were computed it means we need to perform the search
+        if (!_searchState.value().Matches.has_value())
+        {
+            // Before we search, let's wait a bit:
+            // probably the search criteria or the data are still modified.
+            co_await winrt::resume_after(delay);
 
-        Search search(*GetUiaData(), text.c_str(), direction, sensitivity);
-        auto lock = _terminal->LockForWriting();
-        if (search.FindNext())
+            // Switch back to Dispatcher so we can set the Searching status
+            co_await winrt::resume_foreground(Dispatcher());
+            if (auto control{ weakThis.get() })
+            {
+                // If search box was collapsed or the new one search was triggered - let's cancel this one
+                if (!_searchState.has_value() || _searchState.value().SearchId != originalSearchId)
+                {
+                    co_return;
+                }
+
+                // Let's mark the start of searching
+                if (_searchBox)
+                {
+                    _searchBox->SetStatus(_searchState.value().Status());
+                    _searchBox->SetNavigationEnabled(false);
+                }
+
+                std::vector<std::pair<COORD, COORD>> matches;
+                if (!_searchState.value().Text.empty())
+                {
+                    const Search::Sensitivity sensitivity = _searchState.value().CaseSensitive ?
+                                                                Search::Sensitivity::CaseSensitive :
+                                                                Search::Sensitivity::CaseInsensitive;
+
+                    Search search(*GetUiaData(), _searchState.value().Text.c_str(), Search::Direction::Forward, sensitivity);
+                    while (co_await _SearchOne(search))
+                    {
+                        // if search box was collapsed or the new one search was triggered - let's cancel this one
+                        if (!_searchState.has_value() || _searchState.value().SearchId != originalSearchId)
+                        {
+                            co_return;
+                        }
+
+                        matches.push_back(search.GetFoundLocation());
+                    }
+
+                    // if search box was collapsed or the new one search was triggered - let's cancel this one
+                    if (!_searchState.has_value() || _searchState.value().SearchId != originalSearchId)
+                    {
+                        co_return;
+                    }
+                }
+                _searchState.value().Matches.emplace(matches);

```

Comment:
do we need to like, `std::move` the `matches`? I never remember with these STL containers.